blog/_posts/2023-12-26-the-heart-of-a-language-server.adoc

```diff
@@ -0,0 +1,282 @@
+= The Heart of a Language Server
+:sectanchors:
+:page-layout: post
+
+In this post, I want to expand on one curious comment from rust-analyzer code base.
+You can find the comment https://github.com/rust-lang/rust-analyzer/blob/34cffbf1d75fb6b5cb6bc68a9854b20dc74f135d/crates/hir/src/semantics/source_to_def.rs#L3-L4[here].
+
+It describes a curious recursive algorithm that is repeated across different language-server-shaped thing:
+I've seen it implemented for Kotlin and C#, and implemented it myself for Rust.
+
+Here's a seemingly random grab bag of IDE features:
+
+- Go to definition
+- Code completion
+- Run test at the cursor
+- Extract variable
+
+What's common among them all?
+All these features are relative to the _current position_ of the cursor!
+The input is not only the state of the code at a give point in time, but a specific location in the source of a project, like `src/main.rs:90:2`.
+
+And the first thing a language server needs to do for any of the above features is to understand what is located at the given offset, semantically speaking.
+Is it an operator, like `+`?
+Is it a name, like `foo`?
+If it is a name, in what context a name is used --- does it _define_ an entity named `foo` or does it _refer_ to a pre-existing entity?
+If it is a reference, then _what_ entity is referenced?
+What type is it?
+
+The first step here is determining a node in the syntax tree which covers the offset.
+This is relatively straightforward:
+
+[source,rust]
+----
+fn node_at_offset(node: SyntaxNode, offset: u32) -> SyntaxNode {
+    assert!(node.text_range().contains(offset));
+    node.children()
+        .find(|it| it.text_range().contains(offset))
+        .map(|it| node_at_offset(it, offset))
+        .unwrap_or(node)
+}
+----
+
+But the syntax tree by itself doesn't contain enough information to drive IDE features.
+Semantic analysis is required.
+
+But the problem with semantic analysis is that it usually involves several layers of intermediate representations, which are only indirectly related to the syntax tree.
+While the syntax tree is relatively uniform, and it is possible to implement a generic traversal like the one above,
+semantic information is usually stored in a menagerie of ad-hoc data structures: trees, graphs, and plain old hash tables.
+
+Traditional compilers attach source span information to semantic elements, which could look like this:
+
+[source,rust]
```

Comment:
This is broken in the rendered view, but it looks fine, so it's probably a GitHub limitation.